org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/JSEditorAssociationOverride.java

```diff
@@ -0,0 +1,66 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Red Hat Inc. and others.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Victor Rubezhny (Red Hat Inc.) - initial implementation
+ *******************************************************************************/
+package org.eclipse.wildwebdeveloper;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.core.runtime.content.IContentType;
+import org.eclipse.ui.IEditorDescriptor;
+import org.eclipse.ui.IEditorInput;
+import org.eclipse.ui.IEditorRegistry;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.internal.genericeditor.ExtensionBasedTextEditor;
+import org.eclipse.ui.internal.genericeditor.GenericEditorWithContentTypeIcon;
+import org.eclipse.ui.internal.genericeditor.GenericEditorWithIconAssociationOverride;
+
+public class JSEditorAssociationOverride extends GenericEditorWithIconAssociationOverride {
+	private Map<String, IEditorDescriptor> defaultDescriptorMap = new HashMap<>();
+	
+	@Override
+	public IEditorDescriptor overrideDefaultEditor(IEditorInput editorInput, IContentType contentType,
+			IEditorDescriptor editorDescriptor) {
+		return editorInput != null ? overrideDefaultEditor(editorInput.getName(), contentType, editorDescriptor)
+				: editorDescriptor;
+	}
+
+	@Override
+	public IEditorDescriptor overrideDefaultEditor(String fileName, IContentType contentType,
+			IEditorDescriptor editorDescriptor) {
+		return getDefaultEditorDescriptorForFile(editorDescriptor, fileName);
+	}
+
+	// Workaround for https://bugs.eclipse.org/bugs/show_bug.cgi?id=559907
+	private IEditorDescriptor getDefaultEditorDescriptorForFile(IEditorDescriptor defaultDescriptor, String fileName) {
+		if (!ExtensionBasedTextEditor.GENERIC_EDITOR_ID.equals(defaultDescriptor.getId())
+				&& fileName != null && !fileName.isEmpty()) {
+			int index = fileName.lastIndexOf('.');
+			if (index > -1) {
+				String extension = fileName.substring(index);
+				if (".js".equalsIgnoreCase(extension.toLowerCase())) {
+					IEditorRegistry editorReg = PlatformUI.getWorkbench()
+							.getEditorRegistry();
+					
+					IEditorDescriptor genericEditorDescriptor = editorReg.findEditor(ExtensionBasedTextEditor.GENERIC_EDITOR_ID);
+					if (genericEditorDescriptor != null) {
+						if (!defaultDescriptorMap.containsKey(fileName)) {
+							defaultDescriptorMap.put(fileName, new GenericEditorWithContentTypeIcon(fileName, genericEditorDescriptor));
+						}
+						return defaultDescriptorMap.get(fileName);
```

Comment:
I think you can skip the map an just return `new GenericEditorWithContentTypeIcon`.

Reply:
Then it will be a memory leak. We shouldn't create yet another descriptor for an editor if we already have one for a file.

Reply:
The map is the memory leak as there is nothing that ever cleans it.
But you're right, we can avoid instantiating too many `GenericEditorWithContentTypeIcon`. Actually, as the fileName is simply used to resolve the content-type and we already know the interesting content-type here, we can even just instantiate it once and store it as a final field.
```java
public class JSEditorAssociationOverride  .... {
  private final IEditorDescriptor genericEditorWithJSIcon;

  public JSEditorAssociationOverride() {
     IEditorDescriptor genericEditor = PlatformUI.getEditorRegistry().findEditor(ExtensionBasedTextEditor.GENERIC_EDITOR_ID);
     this.genericEditorWithJSIcon = new GenericEditorWithContentTypeIcon("blah.js", genericEditor); // name doesn't matter as long as it match Wild Web Developer JS content-type for icon resolution
```

and then it becomes simpler to implement `overrideDefaultEditor` as a `return wildWebDeveloperJSContentType.equals(contentType) || wildWebDeveloperJSContentType.isAssociatedWith(fileName) ? genericEditorWithJSIcon : super.overrideDefaultEditor(...));`

Reply:
yes, if we're not going to have different icons for JS files we can get rid of map and instantiate the `wildWebDeveloperJSContentType` and  `genericEditorWithJSIcon` in the constructor.